src/ensembl/production/metadata/updater/core.py

```diff
@@ -528,138 +587,336 @@ def get_assembly_sequences(self, species_id, assembly):
-                accession=self.get_meta_single_meta_key(species_id, "assembly.accession"),
-                level=level,
-                name=self.get_meta_single_meta_key(species_id, "assembly.name"),
-                accession_body=accession_body,
-                assembly_default=self.get_meta_single_meta_key(species_id, "assembly.default"),
-                tol_id=tol_id,
-                alt_accession=self.get_meta_single_meta_key(species_id, "assembly.alt_accession"),
-                created=func.now(),
-                assembly_uuid=str(uuid.uuid4()),
-                url_name=self.get_meta_single_meta_key(species_id, "assembly.url_name"),
-                is_reference=is_reference
-            )
-            dataset_factory = DatasetFactory(self.metadata_uri)
-            dataset_type = meta_session.query(DatasetType).filter(DatasetType.name == "assembly").first()
-            (dataset_uuid, assembly_dataset, assembly_dataset_attributes,
-             new_genome_dataset) = dataset_factory.create_dataset(meta_session, None, dataset_source,
-                                                                  dataset_type, attributes, "assembly",
-                                                                  assembly.accession, None,
-                                                                  DatasetStatus.PROCESSED)
-            meta_session.add(assembly)
-            meta_session.add(assembly_dataset)
-            assembly_sequences = self.get_assembly_sequences(species_id, assembly)
-            meta_session.add_all(assembly_sequences)
-
-            meta_session.add_all(assembly_dataset_attributes)
-            return assembly, assembly_dataset, assembly_dataset_attributes, assembly_sequences, dataset_source
-
-    def get_or_new_genebuild(self, species_id, meta_session, source=None, existing=False):
+        # Query core DB once upfront - get names and count together
+        incoming_seq_names = self._get_incoming_sequence_names(species_id)
+        incoming_count = len(incoming_seq_names)
+
+        # Case 1: New assembly accession - Fresh load
+        if not assemblies:
+            return self._create_new_assembly(species_id, meta_session, dataset_source, assembly_accession)
+
+        # Check for force new UUID flag
+        force_new_uuid = self.get_meta_single_meta_key(species_id, "assembly.create_new_uuid")
+
+        # Find assemblies that match on sequence count and names
+        matching_assembly = self._find_matching_assembly(assemblies, incoming_seq_names, incoming_count)
+
+        # Case 2: Found exact match - Attach to existing
+        if matching_assembly is not None:
+            return self._attach_to_existing_assembly(matching_assembly, meta_session, assembly_accession,
+                                                     dataset_source)
+
+        # No exact match found - either error or force new
+        if force_new_uuid == "1" or force_new_uuid == 1:
```

Comment:
```suggestion
        if int(force_new_uuid) == 1:
```
Not sure if this makes sense and you are happy with exceptions being raised when an unexpected content is passed in this parameter.